presto-main/src/test/java/io/prestosql/operator/BenchmarkUnnestOperator.java

```diff
@@ -0,0 +1,424 @@
+            return pages;
+        }
+
+        private static List<Page> createInputPages(int positionsPerPage, List<Type> types, InputGenerator inputGenerator, double primitiveNullsUnnest, double rowNullsUnnest)
+        {
+            ImmutableList.Builder<Page> pages = ImmutableList.builder();
+            int numPages = TOTAL_POSITIONS / positionsPerPage;
+
+            for (int i = 0; i < numPages; i++) {
+                Block[] blocks = new Block[types.size()];
+                blocks[0] = inputGenerator.produceBlock(types.get(0), positionsPerPage, 0.0, 0.0);
+                blocks[1] = inputGenerator.produceBlock(types.get(1), positionsPerPage, primitiveNullsUnnest, rowNullsUnnest);
+                if (blocks.length == 3) {
+                    blocks[2] = inputGenerator.produceBlock(types.get(2), positionsPerPage, 0.0, 0.0);
+                }
+                pages.add(new Page(blocks));
+            }
+
+            return pages.build();
+        }
+    }
+
+    @Benchmark
+    public List<Page> unnest(BenchmarkContext context)
+    {
+        DriverContext driverContext = context.createTaskContext().addPipelineContext(0, true, true, false).addDriverContext();
+        Operator operator = context.getOperatorFactory().createOperator(driverContext);
+
+        Iterator<Page> input = context.getPages().iterator();
+        ImmutableList.Builder<Page> outputPages = ImmutableList.builder();
+
+        boolean finishing = false;
+        for (int loops = 0; !operator.isFinished() && loops < 1_000_000; loops++) {
+            if (operator.needsInput()) {
+                if (input.hasNext()) {
+                    Page inputPage = input.next();
+                    operator.addInput(inputPage);
+                }
+                else if (!finishing) {
+                    operator.finish();
+                    finishing = true;
+                }
+            }
+
+            Page outputPage = operator.getOutput();
+            if (outputPage != null) {
+                outputPages.add(outputPage);
+            }
+        }
+
+        return outputPages.build();
+    }
+
+    /**
+     * Helper class for random input generation.
+     */
+    public static class InputGenerator
+    {
+        private static final int INT_UPPER_LIMIT = 1_000_000_000;
+
+        private Random random = new Random(456);
+
+        // lower and upper bounds on nested structures in the input
+        private final int minNestedCardinality;
+        private final int maxNestedCardinality;
+
+        // upper bound on length of strings generated for VARCHAR elements
+        private final int maxStringLength;
+
+        public InputGenerator(int minNestedCardinality, int maxNestedCardinality, int maxStringLength)
+        {
+            checkArgument(minNestedCardinality >= 0 && maxNestedCardinality >= 0 && minNestedCardinality <= maxNestedCardinality, "cardinality bounds are invalid");
+            checkArgument(maxStringLength >= 0, "invalid string length");
```

Comment:
Be more explicit about what's invalid. Say something like "max string length must be >= 0"